storage/pebble/execution_data_tracker.go

```diff
@@ -0,0 +1,406 @@
+// By appending a suffix of 0xff bytes to the prefix, all keys that start with the given prefix are captured before
+// completing the iteration.
+var ffBytes = bytes.Repeat([]byte{0xFF}, storage.BlobRecordKeyLength)
+
+type StorageOption func(*ExecutionDataTracker)
+
+var _ storage.ExecutionDataTracker = (*ExecutionDataTracker)(nil)
+
+// The ExecutionDataTracker component manages the following information:
+//   - The latest pruned height
+//   - The latest fulfilled height
+//   - The set of CIDs of execution data blobs known at each height, allowing removal of blob data from local storage
+//     once a fulfilled height is pruned
+//   - For each CID, the most recent height at which it was observed, ensuring that blob data needed at higher heights
+//     is not removed during pruning
+//
+// The component invokes the provided prune callback for a CID when the last height at which that CID appears is pruned.
+// This callback can be used to delete the corresponding blob data from the blob store.
+type ExecutionDataTracker struct {
+	// Ensures that pruning operations are not run concurrently with other database writes.
+	// Acquires the read lock for non-prune WRITE operations.
+	// Acquires the write lock for prune WRITE operations.
+	mu sync.RWMutex
+
+	db            *pebble.DB
+	pruneCallback storage.PruneCallback
+	logger        zerolog.Logger
+}
+
+// WithPruneCallback is used to configure the ExecutionDataTracker with a custom prune callback.
+func WithPruneCallback(callback storage.PruneCallback) StorageOption {
+	return func(s *ExecutionDataTracker) {
+		s.pruneCallback = callback
+	}
+}
+
+// NewExecutionDataTracker initializes a new ExecutionDataTracker.
+//
+// Parameters:
+// - logger: The logger for logging tracker operations.
+// - path: The file path for the underlying Pebble database.
+// - startHeight: The initial fulfilled height to be set if no previous fulfilled height is found.
+// - opts: Additional configuration options such as custom prune callbacks.
+//
+// No errors are expected during normal operation.
+func NewExecutionDataTracker(
+	logger zerolog.Logger,
+	path string,
+	startHeight uint64,
+	opts ...StorageOption,
+) (*ExecutionDataTracker, error) {
+	lg := logger.With().Str("module", "tracker_storage").Logger()
+	db, err := pebble.Open(path, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not open db: %w", err)
+	}
+
+	tracker := &ExecutionDataTracker{
+		db:            db,
+		pruneCallback: func(c cid.Cid) error { return nil },
+		logger:        lg,
+	}
+
+	for _, opt := range opts {
+		opt(tracker)
+	}
+
+	lg.Info().Msgf("initialize tracker with start height: %d", startHeight)
+
+	if err := tracker.init(startHeight); err != nil {
+		return nil, fmt.Errorf("failed to initialize tracker: %w", err)
+	}
+
+	lg.Info().Msgf("tracker initialized")
+
+	return tracker, nil
+}
+
+// TODO: move common logic into separate function to avoid duplication of code
```

Comment:
Removed this TODO comment and separated common logic to to avoid duplication of code ([b12dc9b](https://github.com/onflow/flow-go/pull/6277/commits/b12dc9b62b39aba0a9f7a3f94ad029a5a08ca100))